osbuild/util/sbom/spdx.py

```diff
@@ -1,12 +1,89 @@
+import os
 from datetime import datetime
-from typing import List, Union
+from typing import Callable, Dict, List, Optional, Union
 from uuid import uuid4
 
 import osbuild
 import osbuild.util.sbom.model as sbom_model
 import osbuild.util.sbom.spdx2 as spdx2
 
 
+class SpdxLicenseExpressionFactory:
+    """
+    Factory for creating SPDX license expressions from license strings.
+
+    This factory uses the license-expression package to parse license strings and convert them to SPDX license, if
+    possible.
+
+    The factory also keeps track of all extracted licensing information objects that were created during the conversion
+    process. The extracted licensing information objects are stored in a dictionary, where the key is the license
+    reference ID and the value is the ExtractedLicensingInfo object.
+    """
+
+    def __init__(self, license_index_location=None):
+        self._extracted_license_infos: Dict[str, spdx2.ExtractedLicensingInfo] = {}
+        self._spdx_licensing = None
+        try:
+            get_spdx_licensing = self._get_spdx_licensing()
+            if license_index_location:
+                self._spdx_licensing = get_spdx_licensing(license_index_location)
+            else:
+                self._spdx_licensing = get_spdx_licensing()
+        except ImportError as e:
+            if license_index_location:
+                raise ValueError("The license-expression package is not available. "
+                                 "Specify the license index location has no effect.") from e
+
+    @staticmethod
+    def _get_spdx_licensing() -> Callable:
+        """
+        Return the get_spdx_licensing function from the license-expression package.
+        This is a helper function to make testing and mocking easier.
+        """
+        # pylint: disable=import-outside-toplevel
+        from license_expression import get_spdx_licensing
+        return get_spdx_licensing
+
+    def _to_extracted_license_info(self, license_str: str) -> spdx2.ExtractedLicensingInfo:
+        eli = spdx2.ExtractedLicensingInfo(license_str)
```

Comment:
In the end, I didn't set the `name` argument, because the `license_str` may get really long, wich would result in even longer License Ref ID. So we would need to limit the length of the `name` value to some reasonable length. I was not sure which magic number to use, so I'm open to suggestions. I think that `LicenseRef-<name>-<sha256>` is nicer for humans than `LicenseRef-<sha256>`...

Reply:
I don't like the `_str` postfix to the variable name, let's drop it.

Reply:
I'm sorry, but `license` is a python builtin name...
```
osbuild/util/sbom/spdx.py:42:41: W0622: Redefining built-in 'license' (redefined-builtin)
osbuild/util/sbom/spdx.py:46:40: W0622: Redefining built-in 'license' (redefined-builtin)
```